bluepyemodel/emodel_pipeline/plotting_utils.py

```diff
@@ -0,0 +1,638 @@
+
+def get_simulated_FI_curve_for_plotting(evaluator, responses, prot_name):
+    """Get FI curve data from model used in plotting.
+
+    Args:
+        evaluator (CellEvaluator): cell evaluator
+        responses (dict): responses of the cell model
+        prot_name (str): name of the protocol to use for the FI curve
+    """
+    values = evaluator.fitness_calculator.calculate_values(responses)
+    simulated_freq = []
+    simulated_amp_rel = []
+    simulated_amp = []
+    for val in values:
+        if prot_name.lower() in val.lower():
+            # val is e.g. IV_40.soma.maximum_voltage_from_voltagebase
+            n = val.split(".")
+            # case where protocol has '.' in its name, e.g. IV_40.0
+            if len(n) == 4 and n[1].isdigit():
+                n = [".".join(n[:2]), n[2], n[3]]
+            protocol_name = n[0]
+            amp_temp = float(protocol_name.split("_")[-1])
+            if "mean_frequency" in val:
+                simulated_freq.append(values[val])
+                if "bpo_threshold_current" in responses:
+                    simulated_amp_rel.append(amp_temp)
+                    simulated_amp.append(rel_to_abs_amplitude(amp_temp, responses))
+                else:
+                    simulated_amp_rel.append(numpy.nan)
+                    simulated_amp.append(amp_temp)
+
+    return simulated_amp_rel, simulated_amp, simulated_freq
+
+
+def get_impedance(time, voltage, current, stim_start, stim_end, efel_settings):
+    """Get impedance for plotting.
+
+    Args:
+        time (list): time series
+        voltage (list): voltage series
+        current (list): injected current series
+        stim_start (float): stimulus start time
+        stim_end (float): stimulus end time
+        efel_settings (dict): eFEL settings in the form {setting_name: setting_value}
+    """
+    from scipy.ndimage.filters import gaussian_filter1d
+
+    dt = 0.1
+    Z_max_freq = 50.0
+    if efel_settings is not None:
+        dt = efel_settings.get("interp_step", dt)
+        Z_max_freq = efel_settings.get("impedance_max_freq", Z_max_freq)
+
+    trace = {
+        "T": time,
+        "V": voltage,
+        "I": current,
+        "stim_start": [stim_start],
+        "stim_end": [stim_end],
+    }
+
+    efel.reset()
+    for name, value in efel_settings.items():
+        efel.set_setting(name, value)
+
+    efel_vals = efel.get_feature_values(
+        [trace],
+        [
+            "voltage_base",
+            "steady_state_voltage_stimend",
+            "current_base",
+            "steady_state_current_stimend",
+        ],
+    )
+    if efel_vals[0]["voltage_base"] is not None:
+        holding_voltage = efel_vals[0]["voltage_base"][0]
+    else:
+        holding_voltage = efel_vals[0]["steady_state_voltage_stimend"][0]
+    if efel_vals[0]["current_base"] is not None:
+        holding_current = efel_vals[0]["current_base"][0]
+    else:
+        holding_current = efel_vals[0]["steady_state_current_stimend"][0]
```

Comment:
same here, shouldn't we also check that efel_vals[0]["steady_state_current_stimend"] is not None before accessing it?

Reply:
Also done in latest commit.